### DISCUSSION TITLE
Is the algorithm for Coriolis Matrix applicable to the Modified Denavit-Hartenberg (MDH) representation?

### DISCUSSION THREAD

**Discussion by yanlong658** [ROAM-Lab-ND/spatial_v2_extended#9] [Q&A]
Hello sir,

Currently, I have established my MDH coordinate system as follows, defining the origins of the dynamic parameters on each axis.
![image](https://github.com/ROAM-Lab-ND/spatial_v2_extended/assets/37499691/fca75487-39b3-4690-9f65-1850a1bfd1b3)
Upon setting both angles and angular velocities to zero and using the Coriolis Matrix function, I observed differences in the inertia matrices obtained compared to employing angular accelerations with unit length in the six iterations of Rne. According to my understanding, regardless of whether it is MDH or DH, the algorithm in the paper should be applicable. Is my interpretation correct

Thanks.

**Comment by pwensing** [comment]
Hi there, 

All the algorithms are compatible with MDH — the model data structure is independent of the convention you choose to use.

Regarding your issue, note that if we write the equations of motion as:
$$H(q) \ddot{q} + C(q, \dot{q}) \dot{q} + g(q) = \tau$$
Then the CoriolisMatrix function gives $C(q, \dot{q})$. It sounds like what you are looking for is the joint space inertia matrix $H(q)$ instead. The HandC function should do that for you — it gives two outputs, $H(q)$ and $C(q,\dot{q})\dot{q} +g(q)$.

Best,
-Pat

**Comment by yanlong658** [comment]
Hi Pat,

I indeed need the inertia matrix of the joint.
The following is the MATLAB script file I created for a robotic arm.

========================================================
function robot = VA_model()
% spatial_v2 dynamics library.
a2 = 0.155; 
a3 = 0.600;
a4 = 0.180;
d1 = 0.450;
d4 = 0.637;
d6 = 0.100;
% Modified Denavit Hartenberg parameters alpha_{i-1}, a_{i-1}, and d_i 
% as in J.J.Craig's classic text
mdhTable = [   0     d1       0   ; 
          -pi/2      0      a2   ; 
              0      0      a3   ; 
          -pi/2     d4      a4   ;
           pi/2      0       0   ;
          -pi/2     d6       0  ];


% dhTable = [   0      0       d1   ; 
%           -pi/2     a2       0   ; 
%               0     a3       0   ; 
%           -pi/2     a4      d4   ;
%            pi/2      0       0   ;
%           -pi/2      0      d6  ];

robot.NB = 6;
robot.parent = [0 1 2 3 4 5];
robot.jtype = {'Rz', 'Rz', 'Rz','Rz','Rz','Rz'};
robot.jtype_rotor = {'Rz', 'Rz', 'Rz','Rz','Rz','Rz'};
% robot.has_rotor   = ones(6,1);

robot.gr = { 126,       104,       107.66 ,     89.379    ,    81.77     ,  50  };

robot.Xtree = {  };
for i = 1:6
    alpha = mdhTable(i,1);
    a = mdhTable(i,2);
    d = mdhTable(i,3);
    robot.Xtree{i} = rotx(alpha) * xlt([a 0 0]) * xlt([0 0 d]); % I change it to R(\alpha) * Trnas(a) * Rot(\theta) * Trans(d)
    robot.Xrotor{i} = robot.Xtree{i};
end

robot.gravity = [0 0 0];
for i = 1:robot.NB
    robot.I{i} = eye(6);
    robot.I_rot{i} = eye(6);
end

============================= main Function =============================
clc;clear;
addpath(genpath(pwd))
robot = VA_model();
model = postProcessModel(robot);

% a = [m hx hy hz Ixx Iyy Izz Iyz Ixz Ixy]

a = [1. 1. 1. 1. 1. 1. 1. 0. 0. 0.];
b = [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.];
for i = 1:6
    model.I{i} = inertiaVecToMat(a);
    model.I_rotor{i} = inertiaVecToMat(b);
end
q  = zeros(model.NV, 1);
qd = zeros(model.NV, 1);


[H, C, info] = HandC( model, q, qd);
H

==========================================================================

I use the RNE (Recursive Newton-Euler) algorithm using MDH model six times to calculate the numerical values by setting the joint accelerations to 1.By ( Peter Croke mdh_rne)
The following are the results when setting q, qd, and qdd to zero.
H matrix is:
 
![image](https://github.com/ROAM-Lab-ND/spatial_v2_extended/assets/37499691/d5c634a2-bd27-4fa2-94ff-3ec0fddf325b)


The following are the results when I use the HandC function recommended by you earlier.
H matrix is:
 
![image](https://github.com/ROAM-Lab-ND/spatial_v2_extended/assets/37499691/2ca1bd77-2bc7-4d81-a2c4-b01f077ae369)


I'm not sure if the information I provided is sufficient. If there's anything I haven't filled out correctly, please guide me. 
Thank you for your response, and I wish you a wonderful day.

Best regards,
Long"

**Comment by pwensing** [comment]
Hi Long,

Sorry for my delay -- it is difficult to tell the exact source of your issue, but it doesn't seem to me that the answer you are getting from Corke's package is consistent with the spatial_v2 model you expressed. Specifically, looking at the bottom right entry of the H matrix, the result from the spatial package is giving you what I would expect (the Izz inertia for the last link), while the result from Corke's package is giving you three times that. I am confident that Corke's package is correct, so perhaps the model getting input there is not consistent with the model you've created for spatial_v2. 

Unfortunately, this is not something that I have the bandwidth to help debug. Hopefully, since the original message, you have overcome this hurdle!

Best,
-Pat
